server/src/internalClusterTest/java/org/elasticsearch/cluster/coordination/NodeJoiningIT.java

```diff
@@ -0,0 +1,456 @@
+                            throw new ElasticsearchException("[{}] for [{}] denied", action, connection.getNode());
+                        } else {
+                            connection.sendRequest(requestId, action, request, options);
+                        }
+                    });
+                }
+            }
+
+            // Ensure the logging is as expected
+            try (var mockLog = MockLog.capture(NodeJoinExecutor.class)) {
+
+                // We expect to see a node join message from the new node
+                setNodeJoinMockLog(mockLog, numberOfNodesOriginallyInCluster, masterNode);
+
+                // We haven't changed master nodes yet
+                assertEquals(originalMasterName, internalCluster().getMasterName());
+
+                // Sends a node join request to the original master node. This will fail, and cause a master failover
+                String newNodeName = internalCluster().startDataOnlyNode();
+
+                // Wait until the old master has acknowledged the new master's election
+                safeAwait(previousMasterKnowsNewMasterIsElectedLatch);
+                assertNotEquals(originalMasterName, internalCluster().getMasterName());
+                logger.info("New master is elected");
+
+                mockLog.assertAllExpectationsMatched();
+
+                // Assert the new data node was added
+                ClusterState state = internalCluster().clusterService().state();
+                assertEquals(numberOfNodesOriginallyInCluster + 1, state.nodes().getSize());
+                assertEquals(namesOfDataNodesInOriginalCluster.size() + 1, state.nodes().getDataNodes().size());
+                assertEquals(5, state.nodes().getMasterNodes().size());
+
+                List<String> namesOfDataNodesInNewCluster = getListOfDataNodeNamesFromCluster();
+                assertTrue(namesOfDataNodesInNewCluster.contains(newNodeName));
+                for (String nodeName : namesOfDataNodesInOriginalCluster) {
+                    assertTrue(namesOfDataNodesInNewCluster.contains(nodeName));
+                }
+            }
+        } finally {
+            Releasables.closeExpectNoException(Releasables.wrap(cleanupTasks));
+        }
+    }
+
+    // Tests whether a WARN log is thrown when a node attempts to join a cluster, and then the same master node is re-elected (#126192)
+    @TestLogging(
+        reason = "test includes assertions about logging",
+        value = "org.elasticsearch.cluster.coordination.NodeJoinExecutor:WARN,org.elasticsearch.cluster.coordination.NodeJoinExecutor:INFO"
+    )
+    public void testNodeTriesToJoinClusterAndThenSameMasterIsElected() {
+        final var cleanupTasks = new ArrayList<Releasable>();
+
+        try {
+            ensureSufficientMasterEligibleNodes();
+            DiscoveryNode masterNode = internalCluster().clusterService().state().nodes().getMasterNode();
```

Comment:
Regarding test flakiness, one contributor could be where you get the info that you use in your assertions. Using `internalCluster().clusterService()` doesn't necessarily give you the most up to date value since it might not fetch the instance from the current master node. This is documented on that method, but want to make sure you have consider this random behaviour that this utility can have. You can fetch the master node instance by providing the master node name or similar to the test in `TransportMasterNodeActionIT`, which I assume this is partially based on, use a master client.

Reply:
Having tried that out briefly, I think the test failure is on longer due to the latch you mentioned but rather log expectation assertions. Hope that hleps!

Reply:
(at least based on 20-something runs that used to get 8-10 failures in the latch check, which now only fails once later at the log expectation assertion)